Multi-dimensional Quadrature

pbauman opened the pull request:
The primary purpose of this PR is to deploy an interface for multi-dimensional quadrature and provide an implementation of Monte Carlo and Gaussian quadrature using tensor products of the existing one-dimensional rules. 

There was a small bit of refactoring to introduce `QuadratureBase` from which both 1D and multi-D quadrature rules subclass; this is where the quadrature weights are cached. There was some cleanup of the 1D rules (superfluous functions, using statements). I also moved `InterpolationSurrogateHelper` (and its testing) to `MultidimensionalIndexing` since we can reuse that for the `TensorProductQuadrature`.

The `TensorProductQuadrature` is unit-tested for tensor products of Legendre quadrature for a multi-dimesional linear function. The `MonteCarloQuadrature` unit test is less satisfying - the convergence is so noisy (despite the figure that's on Wikipedia...) that we can't check for a rate of 1/2, so I just check that the rate is negative. I'm open to suggestions on strengthening that test. That said, manually sampling out to 1e6+ samples, I feel confident the implementation is doing the right thing.

pbauman commented:
I noticed I'd accidentally turned off the 1D quadrature testing. I squashed in a fix and pushed here again.

dmcdougall commented on src/misc/src/MonteCarloQuadrature.C:
```diff
@@ -0,0 +1,66 @@
+    UniformVectorRV<V,M> uniform_rv("MonteCarloQuadrature_", //prefix
+                                    this->m_domain);
+
+    const BaseVectorRealizer< V, M > & 	realizer = uniform_rv.realizer();
```
Did you fall asleep on your space bar?

pbauman replied:
HA! Apparently. Sorry, will fix, thanks.

dmcdougall commented on src/misc/inc/MultiDimensionalIndexing.h:
```diff
@@ -22,21 +22,21 @@
 namespace QUESO
 {
-  class InterpolationSurrogateHelper
+  class MultiDimensionalIndexing
```
Are you concerned about this breaking older code?  If so, you might consider a simple `typedef` for a release or two before changing the name.

pbauman replied:
Crap. I didn't think about that really. I don't think so, though, because that code was only used internally in the `InterpolationSurrogate` stuff. In the example, that class is not used at all.

dmcdougall commented on src/misc/inc/1DQuadrature.h:
```diff
@@ -44,8 +45,8 @@ namespace QUESO {
 */
-class
-Base1DQuadrature {
+class Base1DQuadrature : public QuadratureBase
```
There's a name inconsistency here; `Base1DQuadrature` versus `QuadratureBase`.

dmcdougall replied:
Is `Base1DQuadrature` intended to be abstract?  There are no pure virtual functions in here.

dmcdougall replied:
Actually, I notice you didn't add this class, it was already in QUESO.  So adding a pure virtual function would be a breaking change.

pbauman replied:
There used to be a pure virtual function... see c7f2820

pbauman replied:
Do I need to scratch that OCD itch?

pbauman replied:
OK, I'll do it. I guess it's not a big deal.

dmcdougall replied:
I'll leave that decision to you.

dmcdougall replied:
Hah.  Fair enough.

dmcdougall commented on src/misc/src/1DQuadrature.C:
```diff
@@ -33,12 +33,10 @@ Base1DQuadrature::Base1DQuadrature(
-  m_weights       (0)
+  m_order         (order)
 {
   queso_require_less_msg(m_minDomainValue, m_maxDomainValue, "min >= max");
```
Should we check `m_maxDomainValue` is finite?

dmcdougall replied:
Actually it appears in some cases you allow an infinite domain; `GaussianHermite1DQuadrature`.

dmcdougall commented on src/misc/src/MonteCarloQuadrature.C:
```diff
@@ -0,0 +1,66 @@
+        typename QUESO::SharedPtr<V>::Type domain_vec(domain.vectorSpace().newVector());
+
+        // Draw the new point
+        realizer.realization(*domain_vec);
```
This will throw an exception if the volume of the domain isn't finite.  Good.

pbauman replied:
Yeah, I'm not sure how to handle the case when someone wants to use `MonteCarloQuadrature` for integration when one of the dimensions is infinite, e.g. when you're marginalizing a Gaussian pdf.

dmcdougall replied:
You don't use a uniform distribution for the realisations, you use a distribution that isn't improper on an unbounded domain and the weights are divided by PDF values at the points where you have realisations.

Thinking about it, we ought to consider other flavours of Monte Carlo quadrature with various different realiser strategies.  If you know something about the function you're integrating it might influence the choice of realiser.

dmcdougall replied:
https://en.wikipedia.org/wiki/Importance_sampling

pbauman replied:
*facepalm* Of course. Yes, at some point we will probably want to generalize this to do importance sampling.

dmcdougall commented on src/misc/inc/MultiDQuadratureBase.h:
```diff
@@ -0,0 +1,78 @@
+  protected:
+
+    //! Locations of quadrature points
+    /*! \todo We should make this a vector of std::unique_ptr when we go full C++11. */
```
I learned something today.  I always thought that you couldn't fill a `vector` with `unique_ptr`s because the contained type must satisfy the requirements of `CopyConstrubtible` and `CopyAssignable`.  That's true, but it's only true up until C++11, at which point the requirements vary depending on what operations are being applied to the container.

pbauman replied:
Yeah, and it would make sense here since we never intend to transfer ownership away, so someone would have to explicitly invoke `std::move` if they tried to.

pbauman commented on src/misc/inc/QuadratureBase.h:
```diff
@@ -0,0 +1,60 @@
+   * This class contains the weights. Subclasses will have the positions since the position can be one-dimensional
+   * (Base1DQuadrature) or multi-dimensional (MultiDQuadratureBase).
+   */
+  class QuadratureBase
```
Conceptually, it is abstract, but there are no virtual functions to put here. The subclass constructors will populate the weights (in this class) and the positions (in the first layer of subclasses, either 1D or multiD). Basically, this is avoiding putting weights in two different places.

We could do without 1DQuadrature entirely, and adhere to the interface in MultiDQuadratureBase, but since 1DQuadrature was already there, and its interface was specifically 1D, I went the route that I did.

dmcdougall replied:
If you want to prevent clients from instantiating it, make the destructor pure virtual.  I'll leave that detail up to you since it doesn't seem too important.

pbauman replied:
That's a nice idea. I didn't know about pure virtual destructors. I will do that here - I've been saved more than once by the compiler finding when I inadvertently instantiated a class that I wasn't supposed to.

dmcdougall commented:
Merge whenever you're happy @pbauman

pbauman commented:
OK, renamed `QuadratureBase` to `BaseQuadrature`, fixed the spacing issue, made quadrature base classes abstract using pure virtual destructor (TIL), and rebased on dev. Will merge once Travis is happy.
